File: removeDuplicates.cpp
#include<iostream>
#include<stdlib.h>

using namespace std;

struct Node{
	int data;
	struct Node *next;
};

struct Node *newNode(int data){
	Node *temp = new Node;
	temp->data = data;
	temp->next = NULL;
	return temp;
}

void removeDuplicates(struct Node *start){
	struct Node *a;
	struct Node *b;
	struct Node *duplicate;
	a = start;
	
	while(a!=NULL && a->next!=NULL){
		b = a;
		//compare picked element with rest of elements 
		while(b->next != NULL){
			if(a->data == b->next->data){
				duplicate = b->next;
				b->next = b->next->next;
				delete(duplicate);
			}
			else b = b->next;
		}
		a = a->next;
	}
}

void printList(struct Node *node){
	while(node != NULL){
		printf("%d ", node->data);
		node = node->next;
	}
}

int main()
{
    /* The constructed linked list is:
     10->12->11->11->12->11->10*/
    struct Node *start = newNode(10);
    start->next = newNode(12);
    start->next->next = newNode(11);
    start->next->next->next = newNode(11);
    start->next->next->next->next = newNode(12);
    start->next->next->next->next->next = newNode(11);
    start->next->next->next->next->next->next = newNode(10);
 
    printf("Linked list before removing duplicates ");
    printList(start);
 
    removeDuplicates(start);
 
    printf("\nLinked list after removing duplicates ");
    printList(start);
 
    return 0;
}

File: removeDuplicates.java
// Java program to remove duplicates from unsorted 
// linked list
 
class LinkedList {
 
    static Node head;
 
    static class Node {
 
        int data;
        Node next;
 
        Node(int d) {
            data = d;
            next = null;
        }
    }
 
    /* Function to remove duplicates from an
       unsorted linked list */
    void remove_duplicates() {
        Node a = null; Node b = null; Node duplicate = null;
        a = head;
 
        /* Pick elements one by one */
        while (a != null && a.next != null) {
            b = a;
 
            /* Compare the picked element with rest
                of the elements */
            while (b.next != null) {
 
                /* If duplicate then delete it */
                if (a.data == b.next.data) {
 
                    /* sequence of steps is important here */
                    duplicate = b.next;
                    b.next = b.next.next;
                    System.gc();//garbage collection
                } else /* This is tricky */ {
                    b = b.next;
                }
            }
            a = a.next;
        }
    }
 
    void printList(Node node) {
        while (node != null) {
            System.out.print(node.data + " ");
            node = node.next;
        }
    }
 
    public static void main(String[] args) {
        LinkedList list = new LinkedList();
        list.head = new Node(10);
        list.head.next = new Node(12);
        list.head.next.next = new Node(11);
        list.head.next.next.next = new Node(11);
        list.head.next.next.next.next = new Node(12);
        list.head.next.next.next.next.next = new Node(11);
        list.head.next.next.next.next.next.next = new Node(10);
 
        System.out.println("Linked List before removing duplicates : \n ");
        list.printList(head);
 
        list.remove_duplicates();
        System.out.println("");
        System.out.println("Linked List after removing duplicates : \n ");
        list.printList(head);
    }
}

